zndraw/zndraw.py

```diff
@@ -354,31 +359,80 @@ def _modifier_run(self, data) -> None:
-        self.socket.on("scene:trash", self.trash_scene)
+        if isinstance(self.file_io, dict):
+            self.file_io = FileIO(**self.file_io)
+
+        if self.register_socket_events:
+            self.socket.on("webclient:available", self.initialize_webclient)
+            self.socket.on("selection:run", self.selection_run)
+            self.socket.on("analysis:run", self.analysis_run)
+            self.socket.on("upload", self.upload_file)
+            self.socket.on("download:request", self.download_file)
+            self.socket.on("modifier:register", self.register_modifier)
+            self.socket.on("scene:trash", self.trash_scene)
         self._connect()
-        self.socket.wait()
-
-    def initialize_webclient(self, sid):
-        start_time = datetime.datetime.now()
-        with self._set_sid(sid):
-            for idx, atoms in enumerate(self.read_data()):
-                if idx == 0:
-                    self.analysis_schema(atoms)
-                    self.selection_schema()
-                    self.draw_schema()
-                self[idx] = atoms
-                # self.step = idx # double the message count ..., replace with part of the setitem message, benchmark
-        log.warning(f"{datetime.datetime.now() - start_time} Finished sending data.")
+        if self.register_socket_events:
+            self.socket.wait()
+
+    @classmethod
+    def from_self(cls, other):
+        kwargs = dataclasses.asdict(other)
+        kwargs.pop("token")
+        kwargs.pop("_uuid")
+        kwargs["register_socket_events"] = False
+        return cls(**kwargs)
+
+    def initialize_webclient(self, data):
+        print("Initializing new pyclient for webclient")
+        sid = data["sid"]
+        token = data["token"]
+        host = data["host"]
+
+        vis = self.from_self(self)
+
+        if host:
+            start_time = datetime.datetime.now()
+
+            for idx, atoms in enumerate(vis.read_data()):
+                with vis._set_sid(sid):
+                    if idx == 0:
+                        vis.analysis_schema(atoms)
+                        vis.selection_schema()
+                        vis.draw_schema()
+                    vis[idx] = atoms
+                    # self.step = idx # double the message count ..., replace with part of the setitem message, benchmark
+            log.warning(
+                f"{datetime.datetime.now() - start_time} Finished sending data."
+            )
+        else:
+            # load the data from the room host webclient (e.g. token)
+            with vis._set_token(token):
+                n_atoms = len(vis)
+
+            for idx in tqdm.trange(n_atoms, ncols=100):
+                with vis._set_token(token):
+                    for _ in range(self.config.retries):
+                        try:
+                            atoms = vis[idx]
+                            break
+                        except Exception:
```

Comment:
We should properly catch disconnect exceptions, not just blanket catch everything

Reply:
I fully agree - but I've seen a variety of different exceptions here, which I did not fully understand so I've decided to catch them all and reconnect. I'd convert this into an issue which needs further research.